File: packages/core/src/connectionValidator.ts
import type { GraphDSL } from '@ygrecon/dsl';
import type { ValidationIssue } from './types.js';
import { nodeTypeRegistry } from './registry.js';

/**
 * Normalize handle ID - remove -target suffix if present, keep base ID
 */
function normalizeHandleId(handleId: string | null | undefined): string | undefined {
  if (!handleId) return undefined;
  // Remove -target suffix if present (e.g., "top-1-target" -> "top-1")
  return handleId.replace(/-target$/, '');
}

/**
 * Extract handle IDs from edge params
 */
function getEdgeHandles(edge: GraphDSL['edges'][0]): {
  sourceHandle?: string;
  targetHandle?: string;
} {
  const sourceHandle = normalizeHandleId(edge.params.sourceHandle as string | undefined);
  const targetHandle = normalizeHandleId(edge.params.targetHandle as string | undefined);
  return { sourceHandle, targetHandle };
}

/**
 * Validate all edge connections in a graph.
 * Performs base-level structural checks and delegates node-type-specific
 * validation to node type plugins.
 *
 * @param dsl - The graph DSL to validate
 * @returns Array of validation issues (empty if all valid)
 */
export function validateConnections(dsl: GraphDSL): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const nodeMap = new Map<string, GraphDSL['nodes'][0]>();
  const edgesByDirection = new Map<string, GraphDSL['edges'][0][]>(); // "from->to" -> edges[]
  const edgesByHandles = new Map<string, GraphDSL['edges'][0]>(); // "from[handleX]->to[handleY]" -> edge

  // Build node map
  for (const node of dsl.nodes) {
    nodeMap.set(node.id, node);
  }

  // Build edge maps and validate base rules
  for (const edge of dsl.edges) {
    // Skip if nodes don't exist (this is checked in compiler.ts)
    if (!nodeMap.has(edge.from) || !nodeMap.has(edge.to)) {
      continue;
    }

    const directionKey = `${edge.from}->${edge.to}`;
    const { sourceHandle, targetHandle } = getEdgeHandles(edge);

    // Rule 1: No duplicate connections in the same direction
    if (!edgesByDirection.has(directionKey)) {
      edgesByDirection.set(directionKey, []);
    }
    const existingInDirection = edgesByDirection.get(directionKey)!;
    
    // Check if this edge is a duplicate (same from->to, regardless of handles)
    const isDuplicate = existingInDirection.some(
      (e) => e.id !== edge.id && e.from === edge.from && e.to === edge.to
    );
    
    if (isDuplicate) {
      issues.push({
        code: 'DUPLICATE_CONNECTION',
        message: `Duplicate connection from ${edge.from} to ${edge.to}. Only one edge allowed per direction.`,
        edgeId: edge.id,
        nodeId: edge.from,
      });
    } else {
      existingInDirection.push(edge);
    }

    // Rule 2: No reverse connection using the same handles
    if (sourceHandle && targetHandle) {
      const handleKey = `${edge.from}[${sourceHandle}]->${edge.to}[${targetHandle}]`;
      const reverseHandleKey = `${edge.to}[${targetHandle}]->${edge.from}[${sourceHandle}]`;
      
      if (edgesByHandles.has(reverseHandleKey)) {
        const reverseEdge = edgesByHandles.get(reverseHandleKey)!;
        issues.push({
          code: 'REVERSE_CONNECTION_ON_SAME_HANDLES',
          message: `Cannot create reverse connection ${edge.from}[${sourceHandle}]->${edge.to}[${targetHandle}] because reverse connection ${reverseEdge.id} already exists on the same handles.`,
          edgeId: edge.id,
          nodeId: edge.from,
        });
      } else {
        edgesByHandles.set(handleKey, edge);
      }
    }
  }

  // Delegate to node-type-specific validation
  for (const edge of dsl.edges) {
    if (!nodeMap.has(edge.from) || !nodeMap.has(edge.to)) {
      continue;
    }

    const fromNode = nodeMap.get(edge.from)!;
    const toNode = nodeMap.get(edge.to)!;
    const { sourceHandle, targetHandle } = getEdgeHandles(edge);

    // Validate from node's perspective
    const fromNodeType = nodeTypeRegistry.get(fromNode.type);
    if (fromNodeType?.validateConnection) {
      const nodeIssues = fromNodeType.validateConnection(
        dsl,
        edge.from,
        edge.to,
        sourceHandle,
        targetHandle
      );
      issues.push(...nodeIssues.map((issue) => ({ ...issue, edgeId: edge.id })));
    }

    // Validate to node's perspective (for incoming connections)
    const toNodeType = nodeTypeRegistry.get(toNode.type);
    if (toNodeType?.validateConnection) {
      // For incoming validation, we swap from/to to represent the connection from the target's perspective
      const nodeIssues = toNodeType.validateConnection(
        dsl,
        edge.to,
        edge.from,
        targetHandle,
        sourceHandle
      );
      issues.push(...nodeIssues.map((issue) => ({ ...issue, edgeId: edge.id })));
    }
  }

  return issues;
}

File: apps/backend/src/validation/connection_validator.py
"""Connection validation logic for batch validation."""

from ..shared.dsl_schema import ValidationIssue
from ..graph.connection_instance import normalize_handle_id
from ..graph.graph import Graph


def validate_connections(graph: Graph) -> list[ValidationIssue]:
    """
    Validate all connections in a graph.

    Performs base-level structural checks and delegates node-type-specific
    validation to node instance classes.

    Args:
        graph: The Graph object to validate

    Returns:
        List of validation issues (empty if all valid)
    """
    issues: list[ValidationIssue] = []
    node_map = {node.id: node for node in graph.nodes}

    # Validate each connection
    for connection_instance in graph.connections:
        # Skip if nodes don't exist (this should be checked elsewhere)
        if connection_instance.from_node_id not in node_map or connection_instance.to_node_id not in node_map:
            continue

        # 1. Validate connection structure (duplicates, reverse connections)
        structure_issues = connection_instance.validate_structure(graph)
        issues.extend(structure_issues)

        # 2. Validate node-type-specific connections
        if not structure_issues:  # Only check node types if structure is valid
            from_node_instance = node_map[connection_instance.from_node_id]
            to_node_instance = node_map[connection_instance.to_node_id]

            source_handle = normalize_handle_id(connection_instance.visual.source_handle)
            target_handle = normalize_handle_id(connection_instance.visual.target_handle)

            # Validate from node's perspective (outgoing connection)
            node_issues = from_node_instance.validate_connection(
                graph,
                connection_instance.from_node_id,
                connection_instance.to_node_id,
                source_handle,
                target_handle,
                connection_instance.id,
            )
            issues.extend(node_issues)

            # Validate to node's perspective (incoming connection)
            if not node_issues:  # Only check if previous validation passed
                node_issues = to_node_instance.validate_connection(
                    graph,
                    connection_instance.from_node_id,
                    connection_instance.to_node_id,
                    source_handle,
                    target_handle,
                    connection_instance.id,
                )
                issues.extend(node_issues)

    return issues

